### ISSUE TITLE
Question about erbb_signaling example

### ISSUE THREAD

**Issue by thomassligon** [ICB-DCM/PESTO#126]
I have some a basic question about the PESTO example erbb_signaling.

How does the code determine which type of sensitivity to use?

This is not included in any parameters in mainErbSignaling.m. However, it is implicitly included in logLikelihoodErbBSignaling, which specifies the parameters 
    optionsAmici.sensi = 1;
    optionsAmici.sensi_meth = 'adjoint';
if nargout > 1.
But where is nargout defined? The llh function is called somewhere within PESTO, but how does PESTO know which sensitivity to use?

By the way, the PESTO documentation version 1.0 does not include the word "adjoint", so that doesn't help. And, I have already found all instances of "adjoint" in the example files.

**Comment by paulstapor** [comment]
PESTO itself does not care about the type of sensitivity used: The function logLikelihoodErbBSignaling makes this choice, but it is an example for an objective function (=log-likleihood) which is to be provided by the user.

The variable nargout is the number of output arguments with which the log-likelihood function is called. So the loglikelihood function knows it it should compute sensitivities or not. But it will always use adjoints here, since this is by far more efficient for this example.

**Comment by FFroehlich** [comment]
`nargout` is a MATLAB builtin function: https://de.mathworks.com/help/matlab/ref/nargout.html

**Comment by thomassligon** [comment]
So, the objective function sees that nargout>1, and that way it knows that it is being called with a request for a second output, and that must mean that it needs sensitivities. But how do I know how it will be called by PESTO?. How do I know that I need to write the objective function that way? Now, instead of searching the PESTO documentation for "adjoint", I searched for "objective function", and that explains some of the logic, but I didn't find anything saying which input or output parameters are expected.
Finally, I should explain why I want to understand the details of this example. I am using it as a model for my asthma code. I took a very preliminary version of the model and simulated it in Copasi, and used that to eliminate unbounded species. Then I simulated it (without sensitivities) in AMICI and got exactly the same results. Now I am trying to use PESTO and adjoint sensitivities to do parameter estimation based on some data, and I am getting horrible integration errors, like Inf in xdot. Most likely, I have something wrong with the way I am formatting the data or calling PESTO.

**Comment by FFroehlich** [comment]
The objective function is not part of PESTO and the responsability of the user. The exact requirements should probably be specified in https://github.com/ICB-DCM/PESTO/blob/master/getMultiStarts.m

The implementation checking for nargout is a specific optimization that we typically do to avoid computational overhead from unnecessary computation of the gradient (when nargout<=1, the gradient will not be returned by the function).

**Comment by thomassligon** [comment]
Thanks for the feedback. Now I understand the input format better. In addition, I have found the error in my code, and it is now running without AMICI integration errors.
When I first ran AMICI, I imported the model from SBML and ran amiwrap. Since that takes a long time, I commented it out so I didn't rerun it on every test. Then I created a PESTO test based on the erbb example. I also manually added the line model.param = 'log10'; to the _syms file, but I forgot to rerun amiwrap. As a result, the simulate_ file had a line options_ami.pscale = 'lin';. After rerunning amiwrap, that changed to options_ami.pscale = 'log10';.
It's not surprising that a mixture of 'lin' and 'log10' caused integration errors.
